[Feature] delete results

SimonEntholzer opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Closes https://github.com/ls1intum/Artemis/issues/3412

### Description
<!-- Describe your changes in detail -->
Introduces a way to delete results of submissions. (Deletion of the whole submission is already possible)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Navigate to Course Administration -> Exercise / ExerciseGroups -> Participations / Submissions
2. Now click on the number of Submissions of an participation /
![image](https://user-images.githubusercontent.com/33342534/119223316-fcbf8e00-baf8-11eb-9199-7434533b906a.png)
3. Next to every result there should be a delete button. 
![image](https://user-images.githubusercontent.com/33342534/119277919-99317f80-bc22-11eb-95b1-fc0901286631.png)
4. _Results_ should not be able to be deleted if a complaint is linked to them. Don't use the delete submissions button on the right, as it is not part of this PR. It will delete the submission no matter if it has results or complaints!

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
participation-submissions.component.ts: 
Assessment Resource:
Assessment Service:

willergeorg reviewed (changes requested):
Tested on TS1. I had the problem when using the "DELETE" Button next to a result and confirm the deletion in the occurring pop up window, the window seems to freeze in the deletion process. I waited for about 10 minutes, the loading sign next to Delete (see screenshot) is still moving and nothing changes. Same seems to hold for submissions with a open complaint.

![complaint](https://user-images.githubusercontent.com/80622272/119234825-2516af00-bb30-11eb-9cb2-ee4f5c48b376.png)

real-dmy reviewed (commented):
Tested on TS1. I was able to delete the submission even though I had filed a complaint for it. The complaint was also deleted with the submission.
![image](https://user-images.githubusercontent.com/80687511/119264392-4a182a00-bbe3-11eb-8909-5a31a8e5fe45.png)
![image](https://user-images.githubusercontent.com/80687511/119264398-500e0b00-bbe3-11eb-9e50-6bf6e3384893.png)

SimonEntholzer commented:
> Tested on TS1. I was able to delete the submission even though I had filed a complaint for it. The complaint was also deleted with the submission.
> ![image](https://user-images.githubusercontent.com/80687511/119264392-4a182a00-bbe3-11eb-8909-5a31a8e5fe45.png)
> ![image](https://user-images.githubusercontent.com/80687511/119264398-500e0b00-bbe3-11eb-9e50-6bf6e3384893.png)

The deletion of the submission is unchanged, which means it can be deleted even if there is a complaint of one of the results.
I only added the deletion of the _**results**_ of a submission.
![image](https://user-images.githubusercontent.com/33342534/119277891-78692a00-bc22-11eb-82b1-782b455f692f.png)
I added it to the description. :)

SimonEntholzer commented:
> Tested on TS1. I had the problem when using the "DELETE" Button next to a result and confirm the deletion in the occurring pop up window, the window seems to freeze in the deletion process. I waited for about 10 minutes, the loading sign next to Delete (see screenshot) is still moving and nothing changes. Same seems to hold for submissions with a open complaint.
> 
> ![complaint](https://user-images.githubusercontent.com/80622272/119234825-2516af00-bb30-11eb-9cb2-ee4f5c48b376.png)
Should be fixed now, I overlooked one of the switch cases.

argertboja commented:
Tested on TS1 and it works for deleting submissions without complaints, however when I try to delete a submission that has a complaint I receive the following error:
![image](https://user-images.githubusercontent.com/17032718/119281613-a2c4e280-bc36-11eb-9665-da45362fca1a.png)

willergeorg reviewed (changes requested):
Tested on TS2. Deletion of a regular submission works but the same error was displayed that @argertboja already mentioned when I try to delete a submission with an open complaint.

ndricimrr commented:
Tested on TS1. I get same as @argertboja 

ALso, I see the following behavior and want to check if that's right : 
- In a submission of a  Text Exercise, the Delete button shows even if there was no grading done to begin with. As you can see in the images below: 

![delete_shows](https://user-images.githubusercontent.com/18335489/119354317-3da7d600-bca4-11eb-9c6a-60c2ff391231.png)

You are able to reproduce this, when you just open submission for assesment but don't submit anything.

SimonEntholzer commented:
> Tested on TS1 and it works for deleting submissions without complaints, however when I try to delete a submission that has a complaint I receive the following error:
> ![image](https://user-images.githubusercontent.com/17032718/119281613-a2c4e280-bc36-11eb-9665-da45362fca1a.png)

Yea there was a small bug with the text-submissions as the java REST endpoint uses a slightly different URL and I overlooked it. Now everything should work for all exercise types.

SimonEntholzer commented:
> Tested on TS1. I get same as @argertboja
> 
> ALso, I see the following behavior and want to check if that's right :
> 
>     * In a submission of a  Text Exercise, the Delete button shows even if there was no grading done to begin with. As you can see in the images below:
> 
> 
> ![delete_shows](https://user-images.githubusercontent.com/18335489/119354317-3da7d600-bca4-11eb-9c6a-60c2ff391231.png)
> 
> You are able to reproduce this, when you just open submission for assesment but don't submit anything.

When you see: <img src=https://user-images.githubusercontent.com/33342534/119399811-a0679480-bcd9-11eb-9b24-a61051b6832d.png width=100> it means there is a result, but it is not yet submitted. (it is the lock for the assessment)
Since at this point the result already exists (although it has no visible score, as it is has e.g. no completionDate), you should also be able to delete it.

canyelten reviewed (changes requested):
Tested on TS1. Deleting submissions without complaint works. But the following error message is still visible while deleting a submission that has a complaint. 
Http failure response for https://artemistest.ase.in.tum.de/api/text-assessments/text-assessments/text-submissions/437629/delete/131141: 405 OK

ArtemTsurikov reviewed (changes requested):
I still get the error, when trying to delete a modelling exercise with complaints linked to it. (Tested on ts1)
<img width="1274" alt="Screenshot 2021-05-25 at 10 20 04" src="https://user-images.githubusercontent.com/64026687/119464322-e9076800-bd42-11eb-88d4-e99c4025b6c5.png">

SimonEntholzer commented on src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.ts:
```diff
@@ -191,4 +212,47 @@ export class ParticipationSubmissionComponent implements OnInit {
+                case ExerciseType.PROGRAMMING:
+                    this.programmingAssessmentService.deleteAssessment(submission.id, result.id).subscribe(
+                        () => {
+                            submission.results = submission.results?.filter((remainingResult) => remainingResult.id !== result.id);
+                            this.dialogErrorSource.next('');
```
I put it in a function :+1:

SimonEntholzer commented on src/main/webapp/i18n/de/global.json:
```diff
@@ -178,6 +178,7 @@
             "back": "Zurück",
             "cancel": "Abbrechen",
             "delete": "Löschen",
+            "noButtonTextDelete": "",
```
Since I can not change the formatting of it text on the jhiDeleteButton, I need to add the button text manually. To make "delete" not appear twice I need to set it to an empty string here.

SimonEntholzer commented on src/test/javascript/spec/component/participation-submission/participation-submission.component.spec.ts:
```diff
@@ -243,6 +277,126 @@ describe('ParticipationSubmissionComponent', () => {
+            deleteProgrammingAssessmentStub.returns(of({}));
+            findAllSubmissionsOfParticipationStub.returns(of({ body: [submissionWithTwoResults] }));
+            stub(participationService, 'find').returns(of(new HttpResponse({ body: participation1 })));
+        });
```
I removed the async, as it realized it is not needed

SimonEntholzer commented on src/test/javascript/spec/component/participation-submission/participation-submission.component.spec.ts:
```diff
@@ -243,6 +277,126 @@ describe('ParticipationSubmissionComponent', () => {
+            expect(comp.submissions?.length).to.be.equal(1);
+            expect(comp.submissions![0].results?.length).to.be.equal(2);
+            expect(comp.submissions![0].results![0]).to.be.deep.equal(result1);
+        });
```
same here

SimonEntholzer commented:
> I still get the error, when trying to delete a modelling exercise with complaints linked to it. (Tested on ts1)
> <img alt="Screenshot 2021-05-25 at 10 20 04" width="1274" src="https://user-images.githubusercontent.com/64026687/119464322-e9076800-bd42-11eb-88d4-e99c4025b6c5.png">

It is intended that you cannot delete a result if it has a complaint. I changed the error message handling so it looks cleaner, now

katjanakosic reviewed (dismissed):
Worked as expected on TS1. For me, it looked like this: 
![image](https://user-images.githubusercontent.com/64956154/119656602-29421580-be2b-11eb-8fab-96e593cef485.png)

lschlesinger commented on src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:
```diff
@@ -305,4 +305,23 @@ public Result saveManualAssessment(final Submission submission, final List<Feedb
+    /**
+     * Deletes the result of a submission.
+     *
+     * @param submission - the submission which the result belongs to
+     * @param result     - the result that should get deleted
```
nit, maybe I would not use the "-" in oder to be consistent with the other doc strings in the controller; this would apply to all the other doc strings you introduced, so I guess you will keep it as it is :D

FuchsDominik replied:
Agree, also because in other places in this PR the documentation is done without the `-`

lschlesinger commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -78,7 +78,7 @@ export class FileUploadAssessmentComponent implements OnInit, OnDestroy {
         private route: ActivatedRoute,
         private resultService: ResultService,
-        private fileUploadAssessmentsService: FileUploadAssessmentsService,
+        private fileUploadAssessmentsService: FileUploadAssessmentService,
```
nit, I see that the name is sometimes `fileUploadAssessmentService` (without the s), I think it would be good to be consistent here, and you changed the type to the version without "s" to it would fit better anyways.

FuchsDominik replied:
Agree to make it consistent. Did you change every occurance of the `fileUploadAssessmentsService` type?

SimonEntholzer replied:
Ah, yea  refactored the type, but it seems like i forgot to refactor all the places where the variable name with the 's' is used. I'll do that too.

lschlesinger commented on src/main/webapp/app/exercises/shared/participation-submission/participation-submission.component.html:
```diff
@@ -72,6 +72,20 @@ <h4 *ngIf="exercise?.assessmentDueDate">
                             [showGradedBadge]="!isTmpOrSolutionProgrParticipation"
                             [showTestDetails]="isTmpOrSolutionProgrParticipation"
                         ></jhi-result>
+                        <span
+                            class="badge"
+                            style="border-radius: 0px"
+                            jhiDeleteButton
+                            [buttonSize]="buttonSizeSmall"
+                            [actionType]="actionTypeEmpty"
+                            [entityTitle]="'deleteResultButton'"
+                            deleteQuestion="artemisApp.result.delete.question2"
+                            (delete)="deleteResult(result.submission, result)"
+                            [dialogError]="dialogError$"
+                        >
+                            <fa-icon [icon]="'times'"></fa-icon>
+                            {{ 'artemisApp.result.delete.buttonText' | artemisTranslate | uppercase }}
+                        </span>
```
why is this outside of the `jhi-result` whereas the GRADED badge is inside this component. IMO the deletion of a result should be placed in the `jhi-result` component as well. 
Also, why are you using the inline style `border-radius: 0px`? This leads to different styles directly shown next to each other:
![image](https://user-images.githubusercontent.com/41366522/119717658-dd14c680-be66-11eb-8aca-70ae7379d326.png)

SimonEntholzer replied:
So, the different style is intended so it is clear that it is a button, and not just a badge. 

When I started to implement it, I actually thought about putting the delete button in the `jhi-result` as well, but decided against it. `jhi-result` is used in a lot of places, and I only want to show this button on this one page. So I think it makes more sense if the delete functionality is not inside the `jhi-result` itself, as it won't be used almost all of the time. (It keeps the `jhi-result` a bit simpler)

lschlesinger replied:
aaah true! makes perfect sense. maybe somewhat more space between the two of them would underline the different functionality badge vs. button... or not writing uppercased DELETE buttons are not uppercased (such as the delete submission on the right) but badges are? does that make sense?

nicolasruscher replied:
I agree that it should be clear that its a button. Not using uppercase letters would be useful here to make the functionality more clear :)

FuchsDominik replied:
In the end I would say an active verb like "delete" implies that there is the possibility to do an action with this button instead if the adjective "graded" for the badge. Not sure if every use gets that, though :D Not using all uppercase letter might clarify this even more

SimonEntholzer replied:
I removed the uppercase pipe :+1:

lschlesinger reviewed (changes requested):
Functionality works as expected, tested on TS1 with different exercise types with/without complaints 👍 !
Code-wise I added some inline comments.

FuchsDominik commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadAssessmentResource.java:
```diff
@@ -115,6 +115,20 @@ public FileUploadAssessmentResource(AuthorizationCheckService authCheckService,
+     *
+     * @param submissionId - the id of the submission for which the current assessment should be deleted
+     * @param resultId     - the id of the result which should get deleted
+     * @return 200 Ok response if canceling was successful, 403 Forbidden if current user is not the assessor of the submission
```
For my understanding: It seems that we only throw a badRequest error if the submission does not belong to the result, but do we actually check if the user is the assessor? And If so, shouldn't an admin be allowed to delete it as well?

SimonEntholzer replied:
Yeah, I made a mistake in the javadoc there, I changed it.

FuchsDominik commented on src/main/webapp/i18n/de/result.json:
```diff
@@ -10,7 +10,12 @@
             "deleted": "Result gelöscht mit ID {{ param }}",
             "delete": {
-                "question": "Soll Result {{ id }} wirklich dauerhaft gelöscht werden?"
+                "question": "Soll Result {{ id }} wirklich dauerhaft gelöscht werden?",
```
I know you did not change anything there, but shouldn't it be rather called: Soll **Bewertung** {{ id }} wirklich dauerhaft gelöscht werden?

lschlesinger replied:
I agree, and if so, then it should be changed in the whole file shouldn't it?

FuchsDominik replied:
True, I only looked at the changed translations. Seems like it is used all the time in this file. Can you adapt this @SimonEntholzer ? :)

SimonEntholzer replied:
I changed it :)

FuchsDominik reviewed (commented):
In general the code looks indeed good :) Had some small remarks which you could look into

FuchsDominik commented on src/main/webapp/i18n/de/result.json:
```diff
@@ -10,7 +10,12 @@
             "updated": "Result aktualisiert mit ID {{ param }}",
```
```suggestion
            "updated": "Bewertung aktualisiert mit ID {{ param }}",
```

FuchsDominik commented on src/main/webapp/i18n/de/result.json:
```diff
@@ -10,7 +10,12 @@
             "updated": "Result aktualisiert mit ID {{ param }}",
             "deleted": "Result gelöscht mit ID {{ param }}",
```
```suggestion
            "deleted": "Bewertung gelöscht mit ID {{ param }}",
```

FuchsDominik commented on src/main/webapp/i18n/de/result.json:
```diff
@@ -10,7 +10,12 @@
+                }
             },
             "detail": {
                 "title": "Result"
```
```suggestion
                "title": "Bewertung"
```

FuchsDominik reviewed (approved):
The logic itself looks good, would still propose to remove the '-' from the method documentations and the also change the remaining 'Result' translations like mentioned in previous comments

ivanchimeno reviewed (approved):
Works as advertised :D

lschlesinger reviewed (approved):
Re-Approve, still works!

tobiasjungmann reviewed (approved):
Tested in the testing session. Worked as expected!

imstreetspirit reviewed (approved):
tested in testing session, works as described..

FuchsDominik reviewed (approved):
Code-wise approve

coolchock reviewed (approved):
tested on ts3, works as expected

maximilianbergmann reviewed (approved):
Tested on TS3. Works as expected

nicolasruscher reviewed (approved):
looked at the code again, all change request were implemented 👍
